entries: prepend all hierarchies with keyword metadata

cyphar opened the pull request:
In order to ensure that CollectUsedKeywords works on manifests as
expected (it returns the keywords the manifest was generated with),
prepend every manifest with a dummy /set and /unset entry which just
tags the manifest as "this is a keyword used".

Fixes #86.

Signed-off-by: Aleksa Sarai <asarai@suse.de>

cyphar commented:
/ping @vbatts

vbatts commented:
@cyphar it does. I just haven't had the minute to check why the collectkeywords doesn't just read all keywords used, instead of needing something like 
```
# <keywords>
/set size= type= uid= gid= mode= link= nlink= time=
/unset
```
first thing in the manifest.

vbatts commented on walk.go:
```diff
@@ -360,3 +366,38 @@ func signatureEntries(root string) []Entry {
+	return []Entry{
+		{
+			Type: CommentType,
+			Raw:  fmt.Sprintf("#%16s%s", "keywords: ", strings.Join(keywords, ",")),
```
If we do go this route, perhaps this comment could instead be like "# here are the set of keywords used in this manifest"

cyphar replied:
Sure. I was just trying to match the other "metadata" tags. But I can change this to instead just be a plain comment with the list of keywords.

cyphar commented:
@vbatts `CollectUsedKeywords` **does** check all of the keywords in the manifest. The problem is that you don't use all of the keywords on every objects (`sha256digest` isn't used on directories and symlinks for example). So in certain cases you can have manifests that had `-K sha256digest` used to generate them but the keyword was never actually added to the manifest.

However, when generating a new manifest for comparison you can only use the same set of keys used by the old manifest -- so `CollectUsedKeywords` is necessary (but it doesn't tell you what the **intended** set of keywords were). So this is necessary to make sure that `CollectUsedKeywords` gets the set of requested keywords when generating the manifest.

Honestly though this is actually an edge case (it only happens if you have a directory with only symlinks inside it), but the fix is not that bad IMO.

vbatts commented:
Understandable. I'm going to take a peek at how the workflow plays out on
the bsd version too

On Tue, Nov 15, 2016, 02:43 Aleksa Sarai <notifications@github.com> wrote:

> *@cyphar* commented on this pull request.
> ------------------------------
>
> In walk.go <https://github.com/vbatts/go-mtree/pull/87>:
>
> > +
> +// keywordEntries returns a slice of entries that ensure that a manifest
> +// generated with a particular keyword set will still be recognised as having
> +// that keyword set. Namely this is [/set <keywords>, /set <none>].
> +func keywordEntries(keywords []string) []Entry {
> +	// Convert all of the keywords to zero-value keyvals.
> +	kvs := []string{}
> +	for _, kw := range keywords {
> +		kvs = append(kvs, kw+"=")
> +	}
> +
> +	// Create a /set <kvs> and an /unset.
> +	return []Entry{
> +		{
> +			Type: CommentType,
> +			Raw:  fmt.Sprintf("#%16s%s", "keywords: ", strings.Join(keywords, ",")),
>
> Sure. I was just trying to match the other "metadata" tags. But I can
> change this to instead just be a plain comment with the list of keywords.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/vbatts/go-mtree/pull/87>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEF6YnGplwvES0h9y3C5fP5kQ84AKw0ks5q-WKbgaJpZM4KwUcm>
> .
>

vbatts commented:
On  FreeBSD-12.0-CURRENT
```
=== RUN   TestTarCompare
--- FAIL: TestTarCompare (0.06s)
        compare_test.go:446: FAILURE: diff[0] = mtree.InodeDelta{diff:"modified", path:"testdir", new:mtree.Entry{Parent:(*mtree.Entry)(0x30c6c080), Children:[]*mtree.Entry{(*mtree.Entry)(0x30c6c5c0)}, Prev:(*mtree.Entry)(0x30c6c580), Next:(*mtree.Entry)(nil), Set:(*mtree.Entry)(0x30c6c6c0), Pos:16, Raw:"", Name:"testdir", Keywords:[]string{"type=dir", "mode=0755", "tar_time=100.000000000"}, Type:4}, old:mtree.Entry{Parent:(*mtree.Entry)(0x30c706c0), Children:[]*mtree.Entry(nil), Prev:(*mtree.Entry)(0x30c706c0), Next:(*mtree.Entry)(nil), Set:(*mtree.Entry)(0x30c70640), Pos:16, Raw:"", Name:"testdir", Keywords:[]string{"size=36", "type=dir", "mode=0755", "nlink=2", "time=100.987655000"}, Type:4}, keys:[]mtree.KeyDelta{mtree.KeyDelta{diff:"modified", name:"gid", old:"0", new:"1002"}, mtree.KeyDelta{diff:"missing", name:"size", old:"36", new:""}}}
        compare_test.go:446: FAILURE: diff[1] = mtree.InodeDelta{diff:"modified", path:"testdir/anotherfile", new:mtree.Entry{Parent:(*mtree.Entry)(0x30c6c540), Children:[]*mtree.Entry(nil), Prev:(*mtree.Entry)(nil), Next:(*mtree.Entry)(nil), Set:(*mtree.Entry)(0x30c6c6c0), Pos:17, Raw:"", Name:"anotherfile", Keywords:[]string{"size=3", "mode=0644", "tar_time=100.000000000", "sha1digest=7e240de74fb1ed08fa08d38063f6a6a91462a815"}, Type:4}, old:mtree.Entry{Parent:(*mtree.Entry)(0x30c707c0), Children:[]*mtree.Entry(nil), Prev:(*mtree.Entry)(0x30c70740), Next:(*mtree.Entry)(nil), Set:(*mtree.Entry)(0x30c70640), Pos:17, Raw:"", Name:"anotherfile", Keywords:[]string{"size=3", "mode=0644", "time=100.987655000", "sha1digest=7e240de74fb1ed08fa08d38063f6a6a91462a815"}, Type:4}, keys:[]mtree.KeyDelta{mtree.KeyDelta{diff:"modified", name:"gid", old:"0", new:"1002"}}}
        compare_test.go:420: '.' is different in the tar -- this is a bug in the tar generation
        compare_test.go:446: FAILURE: diff[3] = mtree.InodeDelta{diff:"modified", path:"tmpfile", new:mtree.Entry{Parent:(*mtree.Entry)(0x30c6c080), Children:[]*mtree.Entry(nil), Prev:(*mtree.Entry)(nil), Next:(*mtree.Entry)(nil), Set:(*mtree.Entry)(nil), Pos:12, Raw:"", Name:"tmpfile", Keywords:[]string{"size=12", "type=file", "uid=1002", "gid=1002", "mode=0644", "tar_time=100.000000000", "sha1digest=94e66df8cd09d410c62d9e0dc59d3a884e458e05"}, Type:4}, old:mtree.Entry{Parent:(*mtree.Entry)(0x30c706c0), Children:[]*mtree.Entry(nil), Prev:(*mtree.Entry)(nil), Next:(*mtree.Entry)(nil), Set:(*mtree.Entry)(0x30c70640), Pos:13, Raw:"", Name:"tmpfile", Keywords:[]string{"size=12", "mode=0644", "time=100.987655000", "sha1digest=94e66df8cd09d410c62d9e0dc59d3a884e458e05"}, Type:4}, keys:[]mtree.KeyDelta{mtree.KeyDelta{diff:"modified", name:"gid", old:"0", new:"1002"}}}
        compare_test.go:450: expected the diff length to be 0, got 4
```

cyphar commented:
@vbatts I'm not sure why `gid` is different -- that's super odd (there isn't even a `gid` in either of those keyword sets). As for the size, it looks like it's related to the unsupported issue we were talking about before in #83 [though it looks like the **actual** problem is the `gid` one because I think I have code handling the `size` issue].

vbatts commented:
so, the bsd mtree implementation must have some notion of what is "supported" for each `type=`, because the following reports no errors.
```bash
vbatts@beastiepi /tmp/tmp.5ce1xXQh$ ln -s "this is a dummy symlink" link
vbatts@beastiepi /tmp/tmp.5ce1xXQh$ mtree -K link,sha256digest -c -p . > /tmp/m
vbatts@beastiepi /tmp/tmp.5ce1xXQh$ cat /tmp/m
#          user: vbatts                                                                                                                                                                                            
#       machine: beastiepi                                                                                                                                                                                         
#          tree: /tmp/tmp.5ce1xXQh                                                                                                                                                                                 
#          date: Tue Nov 15 16:24:24 2016                                                                                                                                                                          
                                                                                                                                                                                                                   
# .                                                                                                                                                                                                                
/set type=file uid=1002 gid=0 mode=0755 nlink=1 flags=none                                                                                                                                                         
.               type=dir mode=0700 nlink=2 time=1479227060.868937000                                                                                                                                               
    link        type=link time=1479227060.868862000 \                                                                                                                                                              
                link=this\040is\040a\040dummy\040symlink                                                                                                                                                           
vbatts@beastiepi /tmp/tmp.5ce1xXQh$ mtree -K link,sha256digest -f /tmp/m -p . 
vbatts@beastiepi /tmp/tmp.5ce1xXQh$ echo $?
0
```

cyphar commented:
@vbatts #83 and this issue are not related.

#83 is about the fact that `size=...` is not support in tar archives. This means that attempting to compare a directory between a tar and regular manifest will result in errors -- because the tar archive won't be able to generate a `size=...`. This is why we need to be able to specify that `size` is unsupported for tar archives (BSD's `mtree` doesn't have support for tar archives so this isn't an issue for them).

#86 (and this PR) is because of [this line that I added](). Now, you can just remove that check if you want. The reason why I added that is because I wanted to have a more meaningful error message than the one you used to get (where you would get an error about a file being modified because `sha256digest` was missing). And I wanted to make it so that someone trying to verify `-K sha256digest` against a manifest that doesn't have that information won't be confused by the results. As far as I can tell, BSD's `mtree` doesn't have that check (so we can remove it if you want).

The issue is I didn't consider the edge case that `CollectUsedKeywords` will only give you the list of keywords which are actually present in the manifest. So if you have a manifest only full of symlinks then you won't figure out that `sha256digest` was actually used.

It's up to you how you want to solve this. In fact you could actually solve #83 and #86 in similar ways (make it so that `keyword=` means that `keyword` is not supported for that object). I'm not sure it's up to you to be honest.

vbatts commented:
@cyphar so i have modified that check in #92, but unfortunately I got carried away in Typing all the things, so it will likely be a noisy PR to review. It's really only the last commit that fixes #86

vbatts commented:
carrying part of this over to #95

cyphar commented:
@vbatts I just hit this again. This PR still fixes the problem -- with `xattrs` set but no file containing an xattr, the result is that `master` will not return `xattr` if you use `UsedKeywords`.

I'm going to hack around it by forcibly setting the keyword list in `umoci` -- but the fix you added doesn't help at all.
